### PR TITLE
Fix Story History bug

### PR DESCRIPTION
**Pull request by marcosjr182**
There was a bug causing a HistoryView to duplicate itself when actions where triggered by another stories
old webm

* HistoryView was refactored while fixing the bug and should now be able to update itself when an change occurs on its story.

#### After
new webm

### CODE REVIEW

File: app/assets/javascripts/views/history_view.js [hunk 1]
```diff
@@ -5,41 +5,75 @@ module.exports = Backbone.View.extend({
...
+    });
   },
 
   connectionError: function(res) {
     this.$loadingSpin.hide();
+  },
+
+  setHistoryTitle: function() {
+    const $header = this.$el.find('.toggle-title');
+    let title     = this.currentStory.get('title');
```
Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: You don't override the value of `title` so you can use a `const` for this too.

File: app/assets/javascripts/views/history_view.js [hunk 2]
```diff
@@ -5,41 +5,75 @@ module.exports = Backbone.View.extend({
   initialize: function() {
     _.bindAll(this, 'addActivities', 'connectionError');
 
-    this.currentStory = {};
-
-    this.$el.html(this.template({id: 'history', name: I18n.t('projects.show.history')}));
+    this.$el.html(this.template({
+      id: 'history',
+      className:'historycolumn',
```
Comment (halan) [inline app/assets/javascripts/views/history_view.js]: It's ok, there two classes? `historycolumn` and `history_column` ?

File: app/assets/javascripts/views/history_view.js (hunk 1 above)
Comment (marcosjr182) [inline app/assets/javascripts/views/history_view.js]: I may override it when it is a long title, on line 76.

Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: Oh, I didn't notice the inline if, sorry! Can we break the like 76 like this then?

```javascript
if(condition) {
  title = ...;
}
```

It's easier to read.

Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: Also, what 32 means? Can we put it in a variable to make it clearer?

File: app/assets/javascripts/views/history_view.js [hunk 3]
```diff
@@ -5,41 +5,75 @@ module.exports = Backbone.View.extend({
...
 
   connectionError: function(res) {
     this.$loadingSpin.hide();
+  },
+
+  setHistoryTitle: function() {
+    const $header = this.$el.find('.toggle-title');
+    let title     = this.currentStory.get('title');
+
+    if (title.length > 32) title = title.substring(0, 32) + '...';
```
Comment (halan) [inline app/assets/javascripts/views/history_view.js]: let put `32` into a named var.

...or

put this elipses algorithm into a reusable function:

```javascript
const elipses = (str, len = 32, suspension = '...') =>
  str.length > len
    ?  str.substring(0, len) + suspension
    : str
```

and then...

```javascript

let title = elipses(this.currentStory.get('title'));
```

File: app/assets/javascripts/views/history_view.js (hunk 2 above)
Comment (marcosjr182) [inline app/assets/javascripts/views/history_view.js]: yes,
`history_column` works like `epic_column`, `backlog_column`... Those are the `<td />`s containing the column header and the actual column.. 
`historycolumn` opposes to `storycolumn` which is the container holding the stories, and is is present on `ColumnView`s / Backlog, Chilly Bin, In Progress, Done

Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: We should refactor those class names 🤔

Comment (marcosjr182) [inline app/assets/javascripts/views/history_view.js]: Agreed. :confused: what do you guys think of changing `.historycolumn` to `.activitycolumn`? It could match better with its content... but I think it could be more confusing than `.historycolumn`

File: app/assets/javascripts/templates/project_view.ejs [hunk 4]
```diff
@@ -13,7 +13,7 @@
         <td data-column-view="done"></td>
       <% } %>
       <td style="display: none;" data-column-view="epic" data-hideable="false"></td>
-      <td style="display: none;" data-column-view="history" data-hideable="false"></td>
+      <td style="display: none;" data-history-view="history" data-hideable="false"></td>
```
Comment (talyssonoc) [inline app/assets/javascripts/templates/project_view.ejs]: Why is this change necessary? It doesn't make sense for me to have the value `history` in a data-attribute called `data-history`.

File: app/assets/javascripts/views/history_view.js (hunk 2 above)
Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: But why is it necessary to have _two_ classes? I think that's the main question, why isn't it possible to use the currently existent `history_column` for what we want to achieve?

Comment (marcosjr182) [inline app/assets/javascripts/views/history_view.js]: So,  `.storycolumn` and `.historycolumn` have both the same style. However, `.storycolumn` is being used by the ProjectView, to identify columns the columns and instantiate all of them as `ColumnView`s. 

After the changes of this PR, `history_column` is there mainly to match the other columns classes pattern. Every column has a `${name}_column` class, and it is being used on feature tests to assert that the subject being tested is on the correct column. Do you think it would be okay to use `historycolumn`, on those tests related to the History column?

File: app/assets/javascripts/templates/project_view.ejs (hunk 4 above)
Comment (marcosjr182) [inline app/assets/javascripts/templates/project_view.ejs]: Agreed. :) Fixed!

File: app/assets/javascripts/views/history_view.js [hunk 5]
```diff
@@ -5,41 +5,81 @@ module.exports = Backbone.View.extend({
...
+
+  toggle: function() {
+    this.$el.toggle();
+    this.stopListening(this.currentStory, 'sync');
+  },
+
+  handleStoryChange: function(event) {
+    this.setHistoryTitle(event)
+    this.fetchActivities(event)
+    event.stopPropagation();
```
Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: It's a better practice to call `stopPropagation` as the first statement of the method.

File: app/assets/javascripts/views/history_view.js [hunk 6]
```diff
@@ -5,41 +5,81 @@ module.exports = Backbone.View.extend({
...
       date: activity.date,
-      user: this.options.users.get(activity.user_id).attributes.name
+      user: this.options.users.get(activity.user_id).get('name')
     }));
   },
 
   setStory: function(story) {
-    this.$loadingSpin.show();
-    this.currentStory = story.attributes;
+    if (!this.currentStory || story.get('id') !== this.currentStory.get('id')) {
```
Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: What do you think about having a method to this conditional?

File: app/assets/javascripts/views/history_view.js [hunk 7]
```diff
@@ -5,41 +5,81 @@ module.exports = Backbone.View.extend({
...
   connectionError: function(res) {
     this.$loadingSpin.hide();
+  },
+
+  setHistoryTitle: function() {
+    const $header = this.$el.find('.toggle-title');
+    let title     = this.currentStory.get('title');
+
+    title = this.ellipses(title);
+    $header.text(I18n.t('projects.show.history') + " '" + title + "'");
```
Comment (talyssonoc) [inline app/assets/javascripts/views/history_view.js]: You can use template strings here https://developer.mozilla.org/pt-BR/docs/Web/JavaScript/Reference/template_strings
